core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -69,6 +69,7 @@ public void run() {
           // Schedule a shutdown. It fires if we don't receive the ping response within the timeout.
           shutdownFuture = scheduler.schedule(shutdown, keepAliveTimeoutInNanos,
               TimeUnit.NANOSECONDS);
+          pingFuture = null;
```

Comment:
It seems like this should be done in all cases (but doesn't matter with DELAYED). If the state is IDLE, the current code would prevent all further PINGS (because it would never be rescheduled). Tying `pingFuture == null` to when there is no callback scheduled prevents having to consider most of those cases. So move clearing the field to before the if?